### ISSUE TITLE
xmlschema fails to validate keyrefs that refer to LightSourceIDKey

### ISSUE THREAD

**Issue by jmuhlich** [tlambert03/ome-types#13]
The ome schema exposes an implementation deficiency in xmlschema:

* The xpath selector for `LightSourceIDKey` is a union that explicitly enumerates all the extensions of `LightSource` (`Instrument/Laser | Instrument/Arc | ...` to paraphrase)
* The actual definition of the `Instrument` element references `LightSourceGroup` which is an abstract element serving as the SubstitutionGroup head for all the different `LightSource` extensions.
* When xmlschema parses a keyref like `ImagePixelsChannelLightSourceSettingsLightSourceIDKeyRef`, which ensures that a `LightSourceSettings` references an actual `LightSource` that exists in the file, it doesn't understand the link between `Instrument/LightSourceGroup` and all of the extensions. This in turn causes it to see all `LightSourceSettings` as having dangling `LightSource` references and the validation fails.

Only the (few) keyrefs to `LightSourceIDKey` trigger this issue. The only other type with the problematic pattern is `Shape` but there are actually no keyrefs to `ShapeIDKey` -- the keyref is to `ROIIDKey` which is a container for Shapes and is not itself a SubstitutionGroup head.

I don't see a great way to solve this short of getting it fixed in xmlschema. I'm looking into that now but it's been slow going. I've traced it down to `xmlschema.validators.identities.XsdIdentity:build` where it creates `self.elements`, a cache of the schema elements the key's selector can potentially match. The code looks for `Instrument/Laser` etc. which don't directly exist so `.elements` ends up empty, causing the selector to never match anything during validation of an XML document. Maybe we can monkey-patch the `.elements` entry after constructing the `XmlSchema` object? Sounds dicey but I'll take a look.

**Comment by jmuhlich** [comment]
Wow, that actually worked! After building the schema object in `get_schema` I added this (needs better formatting) :
```python
schema.maps.identities['{http://www.openmicroscopy.org/Schemas/OME/2016-06}LightSourceIDKey'].elements = {
    e: None for e in schema.maps.substitution_groups['{http://www.openmicroscopy.org/Schemas/OME/2016-06}LightSourceGroup']
}
```
All of that spelunking in the xmlschema code actually paid off... I'll submit an issue to `xmlschema` to see what the author thinks.

I still need to figure out how to fix the codegen for `Instrument` and the light sources to accept the dict that comes out of this, but at least that's all in our hands now.

**Comment by tlambert03** [comment]
wow.  I'm very impressed!  I'm only barely following the validation failure (due to my very rudimentary understanding of xml schema in the first place).
Which test file(s) were failing due to this?  all of them?

**Comment by jmuhlich** [comment]
I think this makes all the test files pass xmlschema validation. Looks like the problems are all model shortcomings or XML conversion issues around the tricky inheritance-based stuff: light sources, shapes, and annotations.

**Comment by tlambert03** [comment]
great.  Sounds like you're working on it now... but here's one instantiation trick you might consider (if this is really more about pydantic validation).  dataclasses have optional [`InitVars`](https://docs.python.org/3/library/dataclasses.html#init-only-variables) that get passed to `__post_init__` that might be useful here? ... e.g.:
```python
class Instrument:
    ...
    light_source_group: List[LightSourceGroup] = field(default_factory=list)
    laser: InitVar[Optional[Union[Laser, List[Laser]]]] = field(default=None)
    # and all the other concrete options...

    def __post_init__(self, laser, ...):
        if laser:
            lasers = laser if isinstance(laser, list) else [laser]
            light_source_group.extend(lasers)
 
```

**Comment by jmuhlich** [comment]
Oh cool, that may help a lot.

**Comment by tlambert03** [comment]
cool.  and just to complete the example... we know what the names of those concrete members should be from the substitution groups here:
https://github.com/tlambert03/ome-types/blob/8916752394b40e625bd27c0c6a9307239f67c8a7/src/ome_autogen.py#L490-L494

**Comment by jmuhlich** [comment]
However, it would be unfortunate to have all those extra kwargs in the documented user-accessible API. I think we should strive to give the dataclasses the cleanest, most ergonomic API and fix stuff like this in our adapter layer for XML decode/encode.

**Comment by tlambert03** [comment]
is it that you don't want the `laser` attribute to be available (for instance with tab autocomplete) on the instrument instance?  or is that you don't want laser to be present in the signature if `Instrument`?  I actually don't mind the latter... it seems to be how the XML is written after-all.  Is the alternative that the user always has to pre-combine all of the LightSources into a list passed to `light_source_group`?

**Comment by tlambert03** [comment]
> fix stuff like this in our adapter layer for XML decode/encode.

ah ... I guess this implies that you don't image anyone will ever _be_ instantiating an instrument directly with `Instrument(id="Instrument:1", ...)` huh?  but rather that everything will be going in/out through XML?

**Comment by jmuhlich** [comment]
I would love to provide a facade where directly reading and writing on the tree of metadata objects feels Pythonic and users don't need to worry about all the bookkeeping with the subtypes and cross-referencing of IDs and stuff. This means XML decode and encode will need some adaptation, but this is OK.

**Comment by jmuhlich** [comment]
So the guiding philosophy would be that the schema provides very strong guidance about what the dataclass API should look like, but we reserve the right to pythonify bits for the sake of ergonomics.

**Comment by tlambert03** [comment]
Definitely agree there.   I do think pydantic helps a lot there, since people can provide dicts of simple types for the most part and it will take care of the subtyping.  But run with it for a bit and let me know what you come up with 👍

**Comment by jmuhlich** [comment]
Fixed in #16 with some workarounds and hard-coding.
